File: xflow/pipeline.py
from __future__ import annotations

from xflow.agent.rule_based import RuleBasedAgent
from xflow.config import AppConfig
from xflow.fetch import Fetcher, MockFetcher
from xflow.models import Source, Tweet
from xflow.storage import Storage


def build_fetcher(config: AppConfig) -> Fetcher:
    if config.fetch.fetcher != "mock":
        raise ValueError(f"Unsupported fetcher '{config.fetch.fetcher}'. MVP supports only 'mock'.")
    return MockFetcher()


def fetch_source(fetcher: Fetcher, source: Source, default_limit: int) -> list[Tweet]:
    limit = source.limit or default_limit
    if source.source_type == "account":
        return fetcher.fetch_user(source.value, limit)
    if source.source_type == "list":
        return fetcher.fetch_list(source.value, limit)
    if source.source_type == "search":
        return fetcher.search(source.value, limit)
    raise ValueError(f"Unsupported source type: {source.source_type}")


def run_fetch(config: AppConfig, storage: Storage) -> dict[str, int]:
    fetcher = build_fetcher(config)
    agent = RuleBasedAgent(config.agent.keywords, config.agent.push_threshold) if config.agent.enabled else None
    fetched = 0
    analyzed = 0

    for source in config.sources:
        storage.upsert_source(source)
        try:
            tweets = fetch_source(fetcher, source, config.fetch.default_limit)
            for tweet in tweets:
                storage.upsert_tweet(tweet)
                fetched += 1
                if agent:
                    storage.save_analysis(agent.analyze(tweet))
                    analyzed += 1
            storage.save_fetch_state(source, "ok", f"Fetched {len(tweets)} tweets.")
        except Exception as exc:
            storage.save_fetch_state(source, "error", str(exc))
            raise
    return {"fetched": fetched, "analyzed": analyzed, "sources": len(config.sources)}

File: src/pipeline.rs
use crate::agent;
use crate::config::AppConfig;
use crate::fetch::fetch_source;
use crate::storage;
use serde::Serialize;
use sqlx::SqlitePool;
use std::time::Duration;

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct FetchResult {
    pub fetched: i64,
    pub analyzed: i64,
    pub sources: i64,
    pub failed: i64,
    pub errors: Vec<FetchSourceError>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct FetchSourceError {
    pub source_type: String,
    pub source_value: String,
    pub message: String,
}

pub async fn run_fetch(config: &AppConfig, pool: &SqlitePool) -> anyhow::Result<FetchResult> {
    let mut fetched = 0;
    let mut analyzed = 0;
    let mut errors = Vec::new();
    let mut sources = storage::list_sources(pool, true).await?;
    if sources.is_empty() {
        sources = config.parsed_sources();
        storage::ensure_config_sources(pool, &sources).await?;
    }
    for (index, source) in sources.iter().enumerate() {
        if index > 0 {
            delay_between_sources(config).await;
        }
        match fetch_source(config, pool, source).await {
            Ok(tweets) => {
                let source_count = tweets.len() as i64;
                for tweet in tweets {
                    storage::upsert_tweet(pool, &tweet).await?;
                    fetched += 1;
                    if config.agent.enabled {
                        storage::save_analysis(pool, &agent::analyze(&tweet, &config.agent))
                            .await?;
                        analyzed += 1;
                    }
                }
                storage::save_fetch_state(
                    pool,
                    source,
                    "ok",
                    Some(&format!("Fetched {source_count} tweets.")),
                )
                .await?;
            }
            Err(err) => {
                let message = err.to_string();
                storage::save_fetch_state(pool, source, "error", Some(&message)).await?;
                errors.push(FetchSourceError {
                    source_type: source.source_type.as_str().to_string(),
                    source_value: source.value.clone(),
                    message,
                });
            }
        }
    }
    Ok(FetchResult {
        fetched,
        analyzed,
        sources: sources.len() as i64,
        failed: errors.len() as i64,
        errors,
    })
}

async fn delay_between_sources(config: &AppConfig) {
    if config.fetch.fetcher != "x_web" {
        return;
    }
    let min = config.fetch.source_delay_min_seconds;
    let max = config.fetch.source_delay_max_seconds;
    let delay = if max > min {
        min + ((max - min) / 2)
    } else {
        min
    };
    if delay > 0 {
        tokio::time::sleep(Duration::from_secs(delay)).await;
    }
}
